src/object/srv_obj.c

```diff
@@ -3473,10 +3473,23 @@ ds_obj_dtx_handle_one(crt_rpc_t *rpc, struct daos_cpd_sub_head *dcsh,
 			if (rc != 0)
 				goto out;
 
-			rc = vos_obj_punch(ioc->ioc_coc->sc_hdl, dcsr->dcsr_oid,
-				dcsh->dcsh_epoch.oe_value, dth->dth_ver,
-				dcsr->dcsr_api_flags, &dcsr->dcsr_dkey,
-				dcsr->dcsr_nr, dcsr->dcsr_punch.dcp_akeys, dth);
+			if (dcsr->dcsr_opc == DCSO_PUNCH_OBJ)
```

Comment:
could you just do something like

daos_key_t *dkey = dcsr->dcsr_opc == DSCO_PUNCH_OBJ ? NULL : &dcsr->dkey;

and make one call to vos_obj_punch ?